### PR TITLE
Center viewport on zoom

### PR DESCRIPTION
**Pull request by DisherProject**
This should fix issue [#1845](https://github.com/wesnoth/wesnoth/issues/1845).

### CODE REVIEW

**Comment by CelticMinstrel** [comment]
By "should" I assume you mean you tried it and it's fixed now?

**Comment by DisherProject** [comment]
Yes of course :) It worked fine for me

File: src/display.cpp [hunk 1]
```diff
@@ -1993,8 +1993,11 @@ bool display::set_zoom(unsigned int amount, const bool validate_value_and_set_in
 
 	const SDL_Rect& area = map_area();
 
-	xpos_ += (xpos_ + area.w / 2) * amount / zoom_;
-	ypos_ += (ypos_ + area.h / 2) * amount / zoom_;
+	//Turn the zoom factor to a double in order to avoid rounding errors.
+	double zoom_factor = double(new_zoom) / double(zoom_);
+	
+	xpos_ = round_double( ((xpos_ + area.w / 2) * zoom_factor) - (area.w / 2) );
+	ypos_ = round_double( ((ypos_ + area.h / 2) * zoom_factor) - (area.h / 2) );
```
Comment (jyrkive) [inline src/display.cpp]: Please remove the spaces around function parameters:

```
xpos_ = round_double(((xpos_ + area.w / 2) * zoom_factor) - (area.w / 2));
```

**Comment by DisherProject** [comment]
Spaces removed!

**Comment by jyrkive** [comment]
Thanks. Merged. :)

**Comment by jostephd** [comment]
I just created https://github.com/wesnoth/wesnoth/pull/4442/files#diff-85d89e7c81ef538af4d65cee40074b67 with a further tweak to this code to address #1951.

**Pull request merged**